File: documentapi/src/vespa/documentapi/messagebus/routablefactories51.cpp
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "routablefactories51.h"
#include <vespa/documentapi/documentapi.h>
#include <vespa/documentapi/loadtypes/loadtypeset.h>
#include <vespa/document/document.h>
#include <vespa/vespalib/objects/nbostream.h>

using vespalib::nbostream;

namespace documentapi {

bool
RoutableFactories51::DocumentMessageFactory::encode(const mbus::Routable &obj, vespalib::GrowableByteBuffer &out) const
{
    const DocumentMessage &msg = static_cast<const DocumentMessage&>(obj);
    out.putByte(msg.getPriority());
    out.putInt(msg.getLoadType().getId());
    return doEncode(msg, out);
}

mbus::Routable::UP
RoutableFactories51::DocumentMessageFactory::decode(document::ByteBuffer &in,
                                                  const LoadTypeSet& loadTypes) const
{
    uint8_t pri;
    in.getByte(pri);
    uint32_t loadClass = decodeInt(in);

    DocumentMessage::UP msg = doDecode(in);
    if (msg.get() != NULL) {
        msg->setPriority((Priority::Value)pri);
        msg->setLoadType(loadTypes[loadClass]);
    }

    return mbus::Routable::UP(msg.release());
}

bool
RoutableFactories51::DocumentReplyFactory::encode(const mbus::Routable &obj, vespalib::GrowableByteBuffer &out) const
{
    const DocumentReply &msg = static_cast<const DocumentReply&>(obj);
    out.putByte(msg.getPriority());
    return doEncode(msg, out);
}

mbus::Routable::UP
RoutableFactories51::DocumentReplyFactory::decode(document::ByteBuffer &in, const LoadTypeSet&) const
{
    uint8_t pri;
    in.getByte(pri);
    DocumentReply::UP reply = doDecode(in);
    if (reply.get() != NULL) {
        reply->setPriority((Priority::Value)pri);
    }
    return mbus::Routable::UP(reply.release());
}

DocumentMessage::UP
RoutableFactories51::CreateVisitorMessageFactory::doDecode(document::ByteBuffer &buf) const
{
    DocumentMessage::UP ret(new CreateVisitorMessage());
    CreateVisitorMessage &msg = static_cast<CreateVisitorMessage&>(*ret);

    msg.setLibraryName(decodeString(buf));
    msg.setInstanceId(decodeString(buf));
    msg.setControlDestination(decodeString(buf));
    msg.setDataDestination(decodeString(buf));
    msg.setDocumentSelection(decodeString(buf));
    msg.setMaximumPendingReplyCount(decodeInt(buf));

    int32_t len = decodeInt(buf);
    msg.getBuckets().reserve(len);
    for (int32_t i = 0; i < len; i++) {
        int64_t val;
        buf.getLong(val); // NOT using getLongNetwork
        msg.getBuckets().push_back(document::BucketId(val));
    }

    msg.setFromTimestamp(decodeLong(buf));
    msg.setToTimestamp(decodeLong(buf));
    msg.setVisitRemoves(decodeBoolean(buf));
    msg.setFieldSet(decodeString(buf));
    msg.setVisitInconsistentBuckets(decodeBoolean(buf));
    msg.getParameters().deserialize(_repo, buf);
    msg.setVisitorDispatcherVersion(50);
    msg.setVisitorOrdering((document::OrderingSpecification::Order)decodeInt(buf));
    msg.setMaxBucketsPerVisitor(decodeInt(buf));

    return ret;
}

bool
RoutableFactories51::CreateVisitorMessageFactory::doEncode(const DocumentMessage &obj, vespalib::GrowableByteBuffer &buf) const
{
    const CreateVisitorMessage &msg = static_cast<const CreateVisitorMessage&>(obj);

    buf.putString(msg.getLibraryName());
    buf.putString(msg.getInstanceId());
    buf.putString(msg.getControlDestination());
    buf.putString(msg.getDataDestination());
    buf.putString(msg.getDocumentSelection());
    buf.putInt(msg.getMaximumPendingReplyCount());
    buf.putInt(msg.getBuckets().size());

    const std::vector<document::BucketId> &buckets = msg.getBuckets();
    for (std::vector<document::BucketId>::const_iterator it = buckets.begin();
         it != buckets.end(); ++it)
    {
        uint64_t val = it->getRawId();
        buf.putBytes((const char*)&val, 8);
    }

    buf.putLong(msg.getFromTimestamp());
    buf.putLong(msg.getToTimestamp());
    buf.putBoolean(msg.visitRemoves());
    buf.putString(msg.getFieldSet());
    buf.putBoolean(msg.visitInconsistentBuckets());

    int len = msg.getParameters().getSerializedSize();
    char *tmp = buf.allocate(len);
    document::ByteBuffer dbuf(tmp, len);
    msg.getParameters().serialize(dbuf);

    buf.putInt(msg.getVisitorOrdering());
    buf.putInt(msg.getMaxBucketsPerVisitor());

    return true;
}

DocumentMessage::UP
RoutableFactories51::GetDocumentMessageFactory::doDecode(document::ByteBuffer &buf) const
{
    return DocumentMessage::UP(
            new GetDocumentMessage(decodeDocumentId(buf),
                                   decodeString(buf)));
}

bool
RoutableFactories51::GetDocumentMessageFactory::doEncode(const DocumentMessage &obj,
                                                         vespalib::GrowableByteBuffer &buf) const
{
    const GetDocumentMessage &msg = static_cast<const GetDocumentMessage&>(obj);

    encodeDocumentId(msg.getDocumentId(), buf);
    buf.putString(msg.getFieldSet());
    return true;
}

DocumentReply::UP
RoutableFactories51::DocumentIgnoredReplyFactory::doDecode(document::ByteBuffer& buf) const
{
    (void) buf;
    return DocumentReply::UP(new DocumentIgnoredReply());
}

bool
RoutableFactories51::DocumentIgnoredReplyFactory::doEncode(
        const DocumentReply& obj,
        vespalib::GrowableByteBuffer& buf) const
{
    (void) obj;
    (void) buf;
    return true;
}

}

File: documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/RoutableFactories51.java
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.documentapi.messagebus.protocol;

import com.yahoo.document.BucketId;
import com.yahoo.document.DocumentId;
import com.yahoo.document.FixedBucketSpaces;
import com.yahoo.document.serialization.DocumentDeserializer;
import com.yahoo.document.serialization.DocumentSerializer;
import com.yahoo.vespa.objects.Deserializer;

import java.util.Map;

/**
 * This class encapsulates all the {@link RoutableFactory} classes needed to implement serialization for the document
 * protocol. When adding new factories to this class, please KEEP THE THEM ORDERED alphabetically like they are now.
 *
 */
public abstract class RoutableFactories51 extends RoutableFactories50 {

    public static class CreateVisitorMessageFactory extends DocumentMessageFactory {

        protected String decodeBucketSpace(Deserializer deserializer) {
            return FixedBucketSpaces.defaultSpace();
        }

        @Override
        protected DocumentMessage doDecode(DocumentDeserializer buf) {
            CreateVisitorMessage msg = new CreateVisitorMessage();
            msg.setLibraryName(decodeString(buf));
            msg.setInstanceId(decodeString(buf));
            msg.setControlDestination(decodeString(buf));
            msg.setDataDestination(decodeString(buf));
            msg.setDocumentSelection(decodeString(buf));
            msg.setMaxPendingReplyCount(buf.getInt(null));

            int size = buf.getInt(null);
            for (int i = 0; i < size; i++) {
                long reversed = buf.getLong(null);
                long rawid = ((reversed >>> 56) & 0x00000000000000FFl) | ((reversed >>> 40) & 0x000000000000FF00l) |
                             ((reversed >>> 24) & 0x0000000000FF0000l) | ((reversed >>> 8) & 0x00000000FF000000l) |
                             ((reversed << 8) & 0x000000FF00000000l) | ((reversed << 24) & 0x0000FF0000000000l) |
                             ((reversed << 40) & 0x00FF000000000000l) | ((reversed << 56) & 0xFF00000000000000l);
                msg.getBuckets().add(new BucketId(rawid));
            }

            msg.setFromTimestamp(buf.getLong(null));
            msg.setToTimestamp(buf.getLong(null));
            msg.setVisitRemoves(buf.getByte(null) == (byte)1);
            msg.setFieldSet(decodeString(buf));
            msg.setVisitInconsistentBuckets(buf.getByte(null) == (byte)1);

            size = buf.getInt(null);
            for (int i = 0; i < size; i++) {
                String key = decodeString(buf);
                int sz = buf.getInt(null);
                msg.getParameters().put(key, buf.getBytes(null, sz));
            }

            msg.setVisitorOrdering(buf.getInt(null));
            msg.setMaxBucketsPerVisitor(buf.getInt(null));
            msg.setVisitorDispatcherVersion(50);
            msg.setBucketSpace(decodeBucketSpace(buf));
            return msg;
        }

        protected boolean encodeBucketSpace(String bucketSpace, DocumentSerializer buf) {
            return FixedBucketSpaces.defaultSpace().equals(bucketSpace);
        }

        @Override
        protected boolean doEncode(DocumentMessage obj, DocumentSerializer buf) {
            CreateVisitorMessage msg = (CreateVisitorMessage)obj;
            encodeString(msg.getLibraryName(), buf);
            encodeString(msg.getInstanceId(), buf);
            encodeString(msg.getControlDestination(), buf);
            encodeString(msg.getDataDestination(), buf);
            encodeString(msg.getDocumentSelection(), buf);
            buf.putInt(null, msg.getMaxPendingReplyCount());

            buf.putInt(null, msg.getBuckets().size());
            for (BucketId id : msg.getBuckets()) {
                long rawid = id.getRawId();
                long reversed = ((rawid >>> 56) & 0x00000000000000FFl) | ((rawid >>> 40) & 0x000000000000FF00l) |
                                ((rawid >>> 24) & 0x0000000000FF0000l) | ((rawid >>> 8) & 0x00000000FF000000l) |
                                ((rawid << 8) & 0x000000FF00000000l) | ((rawid << 24) & 0x0000FF0000000000l) |
                                ((rawid << 40) & 0x00FF000000000000l) | ((rawid << 56) & 0xFF00000000000000l);
                buf.putLong(null, reversed);
            }

            buf.putLong(null, msg.getFromTimestamp());
            buf.putLong(null, msg.getToTimestamp());
            buf.putByte(null, msg.getVisitRemoves() ? (byte)1 : (byte)0);
            encodeString(msg.getFieldSet(), buf);
            buf.putByte(null, msg.getVisitInconsistentBuckets() ? (byte)1 : (byte)0);

            buf.putInt(null, msg.getParameters().size());
            for (Map.Entry<String, byte[]> pairs : msg.getParameters().entrySet()) {
                encodeString(pairs.getKey(), buf);
                byte[] b = pairs.getValue();
                buf.putInt(null, b.length);
                buf.put(null, b);
            }

            buf.putInt(null, msg.getVisitorOrdering());
            buf.putInt(null, msg.getMaxBucketsPerVisitor());
            return encodeBucketSpace(msg.getBucketSpace(), buf);
        }
    }

    public static class GetDocumentMessageFactory extends DocumentMessageFactory {

        @Override
        protected DocumentMessage doDecode(DocumentDeserializer buf) {
            return new GetDocumentMessage(new DocumentId(buf), decodeString(buf));
        }

        @Override
        protected boolean doEncode(DocumentMessage obj, DocumentSerializer buf) {
            GetDocumentMessage msg = (GetDocumentMessage)obj;
            msg.getDocumentId().serialize(buf);
            encodeString(msg.getFieldSet(), buf);
            return true;
        }
    }

    public static class DocumentIgnoredReplyFactory extends DocumentReplyFactory {
        @Override
        protected DocumentReply doDecode(DocumentDeserializer buf) {
            return new DocumentIgnoredReply();
        }

        @Override
        protected boolean doEncode(DocumentReply obj, DocumentSerializer buf) {
            return true;
        }
    }
}
